x-pack/plugins/cases/server/client/utils.ts

```diff
@@ -443,3 +447,122 @@ export const sortToSnake = (sortField: string | undefined): SortFieldCase => {
       return SortFieldCase.createdAt;
   }
 };
+
+/**
+ * Wraps the Authorization class' ensureAuthorized call in a try/catch to handle the audit logging
+ * on a failure.
+ */
+export async function ensureAuthorized({
+  owner,
+  operation,
+  savedObjectID,
+  authorization,
+  auditLogger,
+}: {
+  owner: string;
+  operation: OperationDetails;
+  savedObjectID: string;
+  authorization: PublicMethodsOf<Authorization>;
+  auditLogger?: AuditLogger;
+}) {
+  try {
+    return await authorization.ensureAuthorized(owner, operation);
+  } catch (error) {
+    auditLogger?.log(createAuditMsg({ operation, error, savedObjectID }));
+    throw error;
+  }
+}
+
+/**
+ * Describes an entity with the necessary fields to identify if the user is authorized to interact with the saved object
+ * returned from some find query.
+ */
+interface OwnerEntity {
+  owner: string;
+  id: string;
+}
+
+/**
+ * Wraps the Authorization class' method for determining which found saved objects the user making the request
+ * is authorized to interact with.
+ */
+export async function getAuthorizationFilter({
+  operation,
+  authorization,
+  auditLogger,
+}: {
+  operation: OperationDetails;
+  authorization: PublicMethodsOf<Authorization>;
+  auditLogger?: AuditLogger;
+}) {
+  try {
+    const {
+      filter,
+      ensureSavedObjectIsAuthorized,
+      logSuccessfulAuthorization,
+    } = await authorization.getFindAuthorizationFilter(operation);
+    return {
+      filter,
+      ensureSavedObjectsAreAuthorized: (entities: OwnerEntity[]) => {
+        for (const entity of entities) {
+          try {
+            ensureSavedObjectIsAuthorized(entity.owner);
+            auditLogger?.log(createAuditMsg({ operation, savedObjectID: entity.id }));
+          } catch (error) {
+            auditLogger?.log(createAuditMsg({ error, operation, savedObjectID: entity.id }));
+          }
+        }
+      },
+      logSuccessfulAuthorization,
+    };
+  } catch (error) {
+    auditLogger?.log(createAuditMsg({ error, operation }));
+    throw error;
+  }
+}
+
+/**
+ * Creates an AuditEvent describing the state of a request.
+ */
+export function createAuditMsg({
+  operation,
+  outcome,
+  error,
+  savedObjectID,
+}: {
+  operation: OperationDetails;
+  savedObjectID?: string;
+  outcome?: EventOutcome;
+  error?: Error;
+}): AuditEvent {
+  const doc =
+    savedObjectID != null
+      ? `${operation.savedObjectType} [id=${savedObjectID}]`
+      : `a ${operation.docType}`;
+  const message = error
```

Comment:
Is this new code? I think we should be creating these strings with i18n.

Reply:
I saw some previous comments on my previous comments. I'm going to open a discuss issue for this.

Reply:
https://github.com/elastic/kibana/issues/97650